subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/CrossProjectConfigurationReportingGradle.kt

```diff
@@ -0,0 +1,346 @@
+            return CrossProjectConfigurationReportingGradle(gradle, referrerProject, parentCrossProjectModelAccess, parentCrossProjectConfigurator)
+        }
+    }
+
+    private
+    fun maybeWrapListener(listener: Any): Any = when (listener) {
+        is ProjectEvaluationListener -> CrossProjectModelAccessProjectEvaluationListener(listener, referrerProject, crossProjectModelAccess)
+        // all the supported listener types other than ProjectEvaluationListener are already reported as configuration cache problems in non-buildSrc builds
+        else -> listener
+    }
+
+    private
+    fun <T> Closure<T>.withCrossProjectModelAccessChecks(): Closure<T> =
+        CrossProjectModelAccessTrackingClosure(this, referrerProject, crossProjectModelAccess)
+
+    private
+    fun Action<in Project>.withCrossProjectModelAccessCheck(): Action<Project> {
+        val originalAction = this@withCrossProjectModelAccessCheck
+        return Action<Project> {
+            val originalProject = this@Action
+            check(originalProject is ProjectInternal) { "Expected the projects in the model to be ProjectInternal" }
+            originalAction.execute(crossProjectModelAccess.access(referrerProject, originalProject))
+        }
+    }
+
+    private
+    fun Action<in Gradle>.withCrossProjectModelGradleAccessCheck(): Action<Gradle> {
+        val originalAction = this@withCrossProjectModelGradleAccessCheck
+        return Action<Gradle> {
+            val originalGradle = this@Action
+            check(originalGradle is GradleInternal) { "Expected the projects in the model to be ProjectInternal" }
+            originalAction.execute(crossProjectModelAccess.gradleInstanceForProject(referrerProject, originalGradle))
+        }
+    }
+
+    private
+    class CrossProjectModelAccessProjectEvaluationListener(
+        private val delegate: ProjectEvaluationListener,
+        private val referrerProject: ProjectInternal,
+        private val crossProjectModelAccess: CrossProjectModelAccess
+    ) : ProjectEvaluationListener {
+        override fun beforeEvaluate(project: Project) {
+            delegate.beforeEvaluate(crossProjectModelAccess.access(referrerProject, project as ProjectInternal))
+        }
+
+        override fun afterEvaluate(project: Project, state: ProjectState) {
+            delegate.afterEvaluate(crossProjectModelAccess.access(referrerProject, project as ProjectInternal), state)
+        }
+
+        override fun equals(other: Any?): Boolean =
+            javaClass == (other as? CrossProjectModelAccessProjectEvaluationListener)?.javaClass &&
+                other.delegate == delegate &&
+                other.referrerProject == referrerProject
+
+        override fun hashCode(): Int = Objects.hash(delegate, referrerProject)
+
+        override fun toString(): String = "CrossProjectModelAccessProjectEvaluationListener($delegate)"
+    }
+
+    // region delegated members
+    override fun getTaskGraph(): TaskExecutionGraphInternal =
```

Comment:
💄 Would it be possible to get rid of these simple delegations by implementing `ProjectEvaluationListener` `by delegate`?

Reply:
I think it's safer to have these delegated implementations written explicitly, as it gives us better control and protects us against accidental implicit overrides if new members are added to the interface.

Reply:
That makes a lot of sense 👍